apps/04TowersOfHanoi.js

```diff
@@ -19,23 +19,59 @@ function printStacks() {
   console.log("c: " + stacks.c);
 }
 
-function movePiece() {
-  // Your code here
-
+function movePiece(startStack, endStack) {
+  var starting = stacks[startStack];
+  var startPiece = starting.pop();
+  var end = stacks[endStack];
+  var ending = null;
+  ending = end.push(startPiece);
 }
 
-function isLegal() {
-  // Your code here
+function isLegal(startStack, endStack) {
+  var starting = stacks[startStack];
+  var end = stacks[endStack];
+  var ending = null;
+  if (starting[starting.length - 1] > end[end.length - 1]) {
+    console.log("invalid move.");
+    return false;
+  } else {
+    return true;
+  }
+}
 
+function verifyInput(startStack, endStack) {
+  if ((startStack === 'a' || startStack === 'b' || startStack === 'c') && (endStack === 'a' || endStack === 'b' || endStack === 'c')) {
+    return true;
+  } else {
+    console.log('Invalid entry');
+    return false;
+  }
 }
 
 function checkForWin() {
-  // Your code here
-
+  if (stacks.b.length === 4 || stacks.c.length === 4) {
+    console.log('you win');
+    stacks = {
+      a: [4, 3, 2, 1],
+      b: [],
+      c: []
+    };
+    return true;
+  } else {
+    return false;
+  }
 }
 
 function towersOfHanoi(startStack, endStack) {
-  // Your code here
+  startStack = startStack.toLowerCase();
+  endStack = endStack.toLowerCase();
+  if (verifyInput(startStack, endStack)) {
+    if (isLegal(startStack, endStack)) {
```

Comment:
Nice extra step here! I think we can clean this up even more by combining the `if`s with a `&&`